### PR TITLE
Parse config override

### LINKED ISSUE #42: Feature: Strict config parsing
we should strictly parse the config to catch things like typos or unsupported options

### PR DESCRIPTION
**Pull request by EtDu**
Ensure that the override config was typed correctly.
Fixes #42

### CODE REVIEW

**Review by kumavis — CHANGES_REQUESTED** [review]
- break up the individual test into many test (can be in its own file iff you want)
- create a test helper fn so you can just call
- move example configs out of util and into where the test is actually happening (fine to just define them inline)
```js
testConfigParser({ ... }, 'should catch typo ...')
```

File: src/index.js [hunk 1]
```diff
@@ -226,4 +228,34 @@ function applySesTransforms(browserify) {
 
   browserify.transform(removeHtmlComment, { global: true })
   browserify.transform(changeImportString, { global: true })
+}
+
+function parseOverrideConfig(configOverride) {
```
Comment (kumavis) [inline src/index.js]: rename to `validateConfig` not "parse"

File: src/index.js [hunk 2]
```diff
@@ -226,4 +228,34 @@ function applySesTransforms(browserify) {
...
+  if (typeof configOverride !== 'object') {
+    throw new Error("LavaMoat - Expected config override to be an object")
+  }
+  
+  if (Object.keys(configOverride)[0] !== 'resources') {
```
Comment (kumavis) [inline src/index.js]: the order is not important here, but we should check that 'resources' exists

File: src/index.js [hunk 3]
```diff
@@ -226,4 +228,34 @@ function applySesTransforms(browserify) {
...
+    throw new Error("LavaMoat - Expected label 'resources' for configuration key")
+  }
+
+  Object.entries(configOverride['resources']).forEach((resource, index) => {
```
Comment (kumavis) [inline src/index.js]: you can destructure here
`forEach(([packageName, packageOptions], index) => {`

**Comment by kumavis** [comment]
nice, good stuff :+1: 
just some factoring and cleanup and its good to go

File: src/index.js [hunk 4]
```diff
@@ -226,4 +228,34 @@ function applySesTransforms(browserify) {
...
   browserify.transform(removeHtmlComment, { global: true })
   browserify.transform(changeImportString, { global: true })
+}
+
+function validateConfig(configOverride) {
+  if (typeof configOverride !== 'object') {
+    throw new Error("LavaMoat - Expected config override to be an object")
+  }
+  
+  if (!'resources' in Object.keys(configOverride)) {
```
Comment (kumavis) [inline src/index.js]: `'resources' in configOverride` or `Object.keys(configOverride).includes('resources')`

File: src/index.js [hunk 5]
```diff
@@ -226,4 +228,34 @@ function applySesTransforms(browserify) {
...
+    throw new Error("LavaMoat - Expected label 'resources' for configuration key")
+  }
+
+  Object.entries(configOverride['resources']).forEach(([packageName, packageOpts], index) => {
+    if (index === 0 && packageName !== '<root>') {
```
Comment (kumavis) [inline src/index.js]: i dont think we need to validate the presence of `<root>`

File: src/index.js [hunk 6]
```diff
@@ -226,4 +228,34 @@ function applySesTransforms(browserify) {
...
+      throw new Error("LavaMoat - Expected package '<root>' as first resource")
+    }
+    const packageOptions = Object.keys(packageOpts)
+    const packageEntries = Object.values(packageOpts)
+    if (!packageOptions.every(packageOpt => packageOpt === 'globals' || packageOpt === 'packages')) {
+      throw new Error("LavaMoat - Unrecognized package options. Expected 'globals' or 'packages'")
+    }
```
Comment (kumavis) [inline src/index.js:251]:
```js
optionsWhitelist = [...]
packageOptions.every(packageOpt => optionsWhitelist.includes(packageOpt))
```

File: src/index.js [hunk 7]
```diff
@@ -226,4 +228,34 @@ function applySesTransforms(browserify) {
...
+
+    packageEntries.forEach((entry) => {
+      Object.values(entry).forEach((value) => {
+        if (value !== true) {
```
Comment (kumavis) [inline src/index.js]: `'write'` is a valid value for globals. it allows packages to write to `global[key]`

File: test/config.js [hunk 8]
```diff
@@ -297,4 +297,248 @@ test('Config edits trigger re-bundle if using watchify', async (t) => {
 
   t.notOk(configFileString.includes('"three": 12345678'))
   t.ok(updatedConfigFileString.includes('"three": 12345678'))
-})
\ No newline at end of file
+})
+
+test('Bundle fails with first configOverride', async (t) => {
```
Comment (kumavis) [inline test/config.js]: improve test titles
- call out what the mistake is in the test title, eg `'invalid "resources" key'`
- "config validation fails"

File: test/config.js [hunk 9]
```diff
@@ -297,4 +297,248 @@ test('Config edits trigger re-bundle if using watchify', async (t) => {
...
+  testConfigValidator(configOverride, config, true, t)
+})
+
+
+
+async function testConfigValidator(configOverride, config, shouldBeValid, t) {
+  try {
+    await createBundleFromRequiresArray([], {
+      config,
+      configOverride: configOverride
```
Comment (kumavis) [inline test/config.js]:
```suggestion
      configOverride,
```

**Review by kumavis — CHANGES_REQUESTED** [review]
Nice, looking clean. Couple more nits

**Review by EtDu — COMMENTED** [review]
Solid

**Review by kumavis — APPROVED** [review]

**Comment by kumavis** [comment]
i suspect there will be a lint error

**Pull request merged**